crates/ruff_server/src/session/settings.rs

```diff
@@ -1,25 +1,452 @@
+    /// Any fields not specified by either are set to their defaults.
+    pub(super) fn with_workspace(
+        workspace_settings: &ClientSettings,
+        global_settings: &ClientSettings,
+    ) -> Self {
+        Self::new_impl(&[workspace_settings, global_settings])
+    }
+
+    /// Resolves global settings only.
+    pub(super) fn global_only(global_settings: &ClientSettings) -> Self {
+        Self::new_impl(&[global_settings])
+    }
+
+    fn new_impl(all_settings: &[&ClientSettings]) -> Self {
         Self {
-            code_action_deferred_edit_resolution: code_action_data_support
-                && code_action_edit_resolution,
+            fix_all: Self::resolve_or(all_settings, |settings| settings.fix_all, true),
+            organize_imports: Self::resolve_or(
+                all_settings,
+                |settings| settings.organize_imports,
+                true,
+            ),
+            lint_enable: Self::resolve_or(
+                all_settings,
+                |settings| settings.lint.as_ref().and_then(|lint| lint.enable),
+                true,
+            ),
+            disable_rule_comment_enable: Self::resolve_or(
+                all_settings,
+                |settings| {
+                    settings
+                        .code_action
+                        .as_ref()
+                        .and_then(|code_action| code_action.disable_rule_comment.as_ref())
+                        .and_then(|disable_rule_comment| disable_rule_comment.enable)
+                },
+                true,
+            ),
+            fix_violation_enable: Self::resolve_or(
+                all_settings,
+                |settings| {
+                    settings
+                        .code_action
+                        .as_ref()
+                        .and_then(|code_action| code_action.fix_violation.as_ref())
+                        .and_then(|fix_violation| fix_violation.enable)
+                },
+                true,
+            ),
+        }
+    }
+
+    fn resolve_or<T>(
+        all_settings: &[&ClientSettings],
+        get: impl Fn(&ClientSettings) -> Option<T>,
+        default: T,
+    ) -> T {
+        all_settings
+            .iter()
+            .map(Deref::deref)
+            .find_map(get)
+            .unwrap_or(default)
+    }
+}
+
+impl Default for InitializationOptions {
+    fn default() -> Self {
+        Self::GlobalOnly { settings: None }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use insta::assert_debug_snapshot;
+
+    use super::*;
+
+    /// Deserializes a JSON file at `resources/test/fixtures` + the path provided.
+    macro_rules! fixture {
+        ($path:expr) => {
+            serde_json::from_str(include_str!(std::concat!(
+                "../../resources/test/fixtures/",
+                $path
+            )))
+            .expect("test fixture JSON should deserialize")
+        };
+    }
+
```

Comment:
Nit: the fact that the macro concatenates the path breaks some IDE functionality. For example, JetBrain IDEs allow me to click on the argument passed to `include_str!` and it opens the file for me. However, that only works when the path is static. That's why I still think not using a macro here is preferred:

```rust
fn deserialize_fixture<T>(content: &str) -> T {
    serde_json::from_str(content).expect("test fixture JSON should deserialize")
}
```

and inside tests you can either write

```rust
let options: InitializationOptions =
            deserialize_fixture(include_str!("../../resources/test/fixtures/settings/vs_code_initialization_options.json"));

// or 
let options: InitializationOptions =
            deserialize_fixture(include_str!("{}"));
```

Reply:
> For example, JetBrain IDEs allow me to click on the argument passed to `include_str!` and it opens the file for me. However, that only works when the path is static.

Neovim also has a similar functionality where you can press a key (`gf`) and it will take you to that file.